File: Algorithm/Sort/insertion_sort.py
arr = [2, 3, 4, 1, 1, 5, 5, 3, 4, 6, 8, 9, 7]

# 삽입 정렬
for i in range(1, len(arr)):
    for j in range(i, 0, -1):
        if arr[j - 1] > arr[j]:
            arr[j - 1], arr[j] = arr[j], arr[j - 1]
print(arr)

# 삽입 정렬 개선
for i in range(1, len(arr)):
    for j in range(i, 0, -1):
        if arr[j - 1] < arr[j]:
            break
        else:
            arr[j - 1], arr[j] = arr[j], arr[j - 1]
print(arr)

# 삽입 정렬 개선 2
for idx in range(1, len(arr)):
    i = idx
    to_insert = arr[idx]
    while i > 0 and arr[i - 1] > to_insert:
        arr[i] = arr[i - 1]
        i -= 1
    arr[i] = to_insert
print(arr)

File: algorithm/sort/insertionSort.dart
List<int> insertionSort(List<int> arr) {
  for (int i = 1; i < arr.length; i++) {
    int idx = i;
    int toInsert = arr[i];

    while (idx > 0 && arr[idx - 1] > toInsert) {
      arr[idx] = arr[idx - 1];
      idx -= 1;
    }

    arr[idx] = toInsert;
  }

  return arr;
}

void main() {
  List<int> arr = [5, 1, 3, 7, 6, 4, 2];
  List<int> sortedArr = insertionSort(arr);
  print(sortedArr);

  List<int> arr2 = [5, 4, 10, 1, 9, 3, 2, 8, 6, 7];
  List<int> sortedArr2 = insertionSort(arr2);
  print(sortedArr2);

  List<int> arr3 = [1];
  List<int> sortedArr3 = insertionSort(arr3);
  print(sortedArr3);
}